### ISSUE TITLE
Security Exception - Unauthorized OpenVPN API Caller

### ISSUE THREAD

**Issue by aneeshATgithub** [schwabe/ics-openvpn#470]
A Security Exception arises While calling IOpenVPNAPIService's startVPN() after clearing allowed external apps from settings tab in ICS Open VPN.

**Comment by schwabe** [comment]
In the app that remotely controlling OpenVPN for Android? 

Well if that is the case that is what that settings button is intended to do.

**Comment by aneeshATgithub** [comment]
ya..remotely controlling OpenVPN for Android

**Comment by schwabe** [comment]
On getting that Exception you should check if you have still access. And if not decided if you request it again from the user.

**Comment by aneeshATgithub** [comment]
while running remoteExample with my mStartUUID, the vpn is not starting and getting the message EXITING|auth-failure,,,,

**Comment by schwabe** [comment]
well yes, the config there has no valid user credentionals, so it connects and then exits.

**Comment by macdonaldj** [comment]
After getting the prompt to confirm permissions, I am getting the error "Unauthorized OpenVPN API Caller". After I get the result_ok from the confirm dialog, I go to register the callback service for status and get the unauthorized error. After the error, I check the allowed apps and see my app in the list. This used to work, and then started happening after changing my application id in the build gradle. I even uninstalled the app and reinstalled. Thoughts? I am on Android 6.0

**Comment by macdonaldj** [comment]
I assume this started happening after the latest update

**Comment by schwabe** [comment]
@macdonaldj Yes, the bug should be fixed in 0.6.64

**Comment by macdonaldj** [comment]
thanks! sorry about the double post, didn't know if closed issues would get reported back to you. Thanks for the great work!

**Comment by d-schmidt** [comment]
I now had a user getting somehow trigger this.

```
Exception java.lang.SecurityException: Unauthorized OpenVPN API Caller
android.os.Parcel.readException (Parcel.java:1620)
android.os.Parcel.readException (Parcel.java:1573)
de.blinkt.openvpn.api.IOpenVPNAPIService$Stub$Proxy.prepareVPNService (IOpenVPNAPIService.java:323)
...VPNImpl.prepOFAPermissions (VPNImpl.java:233)
```

line 233 is `prepareVPNService()`
```
        // Request permission to use the API
        final Intent controlPermission = ofaService.prepare(context.getPackageName());

        if (controlPermission != null) {
            log.d("prepOFAPermissions() requesting control permission");
            listener.onRequiresOFAPermission1();
            return false;
        }

        // request permission for ofa to set vpn
        final Intent vpnPermission = ofaService.prepareVPNService();
```

It seems [ExternalOpenVPNService.java#L121](https://github.com/schwabe/ics-openvpn/blob/master/main/src/main/java/de/blinkt/openvpn/api/ExternalOpenVPNService.java#L121) failed which is practically impossible. I can't see a bug or a way to improve this code so I guess I just have to ignore the user.

Samsung SM-J510FN with Android 6.0.1 and OfA 0.6.73

**Comment by schwabe** [comment]
I agree that line should not be able to fail @d-schmid double check that openvn for android is really an up date version and line 121 is not from an old version and something else.

**Comment by d-schmidt** [comment]
Extracting the version of OfA was one if the first things I added after the first errors came in. The user uses 0.6.73
I found line 121 going through the source here. It doesn't appear in the stacktrace. On my end in my app I only see the SecurityException using the code above. The only possible reason for `checkOpenVPNPermission()` to fail and `prepare()` to work has to be a uid mismatch. Or some weird preferences bug.
